staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval_test.go

```diff
@@ -0,0 +1,496 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+func intervalFromEvents(events []*watchCacheEvent) *watchCacheInterval {
+	startIndex, endIndex, locker := 0, len(events), &sync.Mutex{}
+	indexer := func(i int) *watchCacheEvent {
+		if len(events) == 0 {
+			return nil
+		}
+		return events[i]
+	}
+	indexValidator := func(_ int) bool { return true }
+
+	return newCacheInterval(startIndex, endIndex, indexer, indexValidator, locker)
+}
+
+func bufferFromEvents(events []*watchCacheEvent) *watchCacheIntervalBuffer {
+	wcib := &watchCacheIntervalBuffer{
+		buffer:     make([]*watchCacheEvent, bufferSize),
+		startIndex: 0,
+		endIndex:   len(events),
+	}
+	copy(wcib.buffer, events)
+
+	return wcib
+}
+
+func generateEvents(start, end int) []*watchCacheEvent {
+	n := end - start
+	events := make([]*watchCacheEvent, n)
+	for i := 0; i < n; i++ {
+		events[i] = &watchCacheEvent{
+			Type: watch.Added,
+			Object: makeTestPod(
+				fmt.Sprintf("pod%d", start+i+1),
+				uint64(start+i+1),
+			),
+		}
+	}
+	return events
+}
+
+func verifyEvent(event, expectedEvent *watchCacheEvent, received, expected int, err error) error {
+	if err != nil {
+		return err
+	}
+
+	if event == nil {
```

Comment:
Yeah, it's not technically needed. The reason I kept it is mainly two things:
- We want to make sure that if no error is returned (`nil`), we always get a non-nil event when we expect events to exist (`received < expected`) - this is more of a sanity check
- When `reflect.DeepEqual` is called, we can pass it a nil event (in case somehow we get `err = nil` and `event = nil`), but in all other cases, if the passed events don't match, then I'd like to log their de-referenced forms and not just pointer values